### DISCUSSION TITLE
Making programs with native nushell output

### DISCUSSION THREAD

**Discussion by TokenRafael** [nushell/nushell#8449] [Ideas]
Sorry if I make something wrong, I am kind of new on this Open Source Collaboration stuff. If I do something wrong please do warn me.

In some of my projects, I was thinking about having nushell as a main tool (building an OS maybe) and for that I would like to integrate some programs, either done or made from 0 by me and some colleges, with nushell, specially when considering input and output.

Doing this in Bash is straight forward: input and output are both strings. But in nushell, all different kinds of data can be used and it would be great to use those instead of getting a string then splitting it in its components so I can finally put it on data structures and process in my program.

Which would be the best way to make an entire toolkit of programs outside nushell, but that can interpret IO in different data structures present in it? Would it be interesting to make, for exemple, a Rust library that gets those inputs and provides them in the program? Or would it just be better to wrap one's program in a nushell script that parses the strings, like interceptors, for the IO?

Personally, I would find it really interesting to do the library option, but I'm interested in knowing why would this be a bad idea.

**Comment by fdncred** [comment]
We're currently working on a standard library. You can see it here https://github.com/nushell/nushell/tree/main/crates/nu-utils/standard_library. This is an entire script based approach to augmenting nushell. This uses modules. We also have overlays. You can look up both those terms in our documentation site nushell.sh.

For an example of how to integrate rust code and make your own thing, I'd use the couchbase shell model. https://github.com/couchbaselabs/couchbase-shell/ They have an entire database cli written with nushell.

**Reply by TokenRafael** [reply]
Nice! Thank you, and sorry if I didn't research enough. This is very helpful!
